### ISSUE TITLE
Panzer: Transition to use DynRankView refactor version of Intrepid2

### ISSUE THREAD

**Issue by rppawlo** [trilinos/Trilinos#274]
In the refactor of Intrepid2, we will stop passing MDFields into Intrepid and instead use Kokkos::Views and Kokkos::DynRankViews.

This is close to being finished.

Issues to complete:
#808 [done]
#809 [done]
#810 [done]
Fix the ResponseScatterEvalautor_Probe (Intrepid2::checkPointwiseInclusion was changed). [deferred - will be added as separate ticket]

Status:
1. Panzer is completed
2. Application codes completed

**Comment by rppawlo** [comment]
@trilinos/panzer @mperego @kyungjoo-kim @bathmatt 

Adding people to this ticket. Looks like we are almost ready to start testing this out.

**Comment by rppawlo** [comment]
Saving some comments from a recent email thread:

Let us have a look as well into this. So far we've only been working on a subset of functions, you'll find in intrepid2/src-kokkos/, and they are enabled by " -D Tpetra_ENABLE_Kokkos_Refactor:BOOL=ON". 

The code passes some unit tests. You can compile the code with the
refactor flag and see how much it is done. Those unit tests also
demonstrate how dynrankview is interfaced. Since all intrepid compoenents
are necessary to get panzer work, you still need to wait panzer
integration a few more days. Our plan does not include tet in this month.
We planned to convert line, quad, hex for hgrad, hdiv, and hcurl. You
probably need to temporary disable tests for tet elements this time. Is
this okay ?

**Comment by rppawlo** [comment]
Generating a list of issues:
1. Changes to intrepid2 are non-backwards compatible. Need to decide on transition strategy - branch or ifdefs.
2. Phalanx: remove support for intrepid2 overloads for MDFields. Remove optional dependence on Intrepid2. Libraries are now completely decoupled! This is a great change.
3. DOF Manager uses Intrepid::FieldContainer objects in many APIs. Need to decide on what to change these to: MDFields or DynRankView objects. MDFields are a little more painful to allocate (on purpose - meant to be handled internally by FieldManager). DynRankView are templated on device type and this is would force us to template quite a bit more code including, for example, the panzer Intrepid2FieldPattern class.

**Comment by rppawlo** [comment]
Just did a grep on FieldContainer to get an idea of how much it is used in panzer. It came back with 562 hits. Yikes! Will attach the file shortly.

**Comment by rppawlo** [comment]
dof manager currently requires phalanx for the node template type. Will remove as part of this ticket. Will add an execution space to panzer and a check such that if phalanx support is enabled in panzer then the default execution spaces must match.

**Comment by rppawlo** [comment]
[panzer_use_of_field_container.txt](https://github.com/trilinos/Trilinos/files/246183/panzer_use_of_field_container.txt)

**Comment by rppawlo** [comment]
Here is the refactor script to convert all FieldContainers to DynRankViews in panzer. Note that this does not protect backwards compatibility. Will have to figure that out later. just trying to get something running with new intrepid2 work.

[refactor2.sh.txt](https://github.com/trilinos/Trilinos/files/246192/refactor2.sh.txt)

**Comment by rppawlo** [comment]
@kyungjoo-kim and @mperego - I ran a replacement script to swap out the FieldContainer code for DynRankView in panzer. But now I get the following compile errors below. It looks like there are some missing defines required in the basis files. Maybe just missing an include header?

If you would like to try this out, I just pushed this code to the branch:
 
panzer_dyn_rank_view_refactor

in my fork of Trilinos on github:

rppawlo/Trilinos

Roger

```
[ 39%] Building CXX object packages/panzer/dof-mgr/src/CMakeFiles/panzer-dof-mgr.dir/Panzer_NodalFieldPattern.cpp.o
In file included from /home/rppawlo/Trilinos/packages/panzer/dof-mgr/src/Panzer_NodalFieldPattern.cpp:45:0:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Cell/Intrepid2_CellTools.hpp:1403:5: warning: multi-line comment [-Wcomment]
     //                / \
     ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Cell/Intrepid2_CellTools.hpp:1407:5: warning: multi-line comment [-Wcomment]
     //            / \     / \
     ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Cell/Intrepid2_CellTools.hpp:1409:5: warning: multi-line comment [-Wcomment]
     //          /      |      \
     ^
In file included from /home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEM.hpp:211:0,
                 from /home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Cell/Intrepid2_CellTools.hpp:66,
                 from /home/rppawlo/Trilinos/packages/panzer/dof-mgr/src/Panzer_NodalFieldPattern.cpp:45:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp: In constructor ‘Intrepid2::Basis_HGRAD_LINE_C1_FEM<ExecSpaceType, outputValueType, pointValueType>::Basis_HGRAD_LINE_C1_FEM()’:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:122:7: error: ‘ordinal_type_array_1d_host’ was not declared in this scope
       ordinal_type_array_1d_host tagView(&tags[0], 8);
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:126:7: error: ‘ordinal_type_array_2d_host’ was not declared in this scope
       ordinal_type_array_2d_host ordinalToTag;
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:127:7: error: ‘ordinal_type_array_3d_host’ was not declared in this scope
       ordinal_type_array_3d_host tagToOrdinal;
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:128:31: error: ‘tagToOrdinal’ was not declared in this scope
       this->setOrdinalTagData(tagToOrdinal,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:129:31: error: ‘ordinalToTag’ was not declared in this scope
                               ordinalToTag,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:130:31: error: ‘tagView’ was not declared in this scope
                               tagView,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:146:17: error: ‘basisCardinality_’ was not declared in this scope
       dofCoords(basisCardinality_,basisCellTopology_.getDimension()); 
                 ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_LINE_C1_FEMDef.hpp:146:35: error: ‘basisCellTopology_’ was not declared in this scope
       dofCoords(basisCardinality_,basisCellTopology_.getDimension()); 
                                   ^
In file included from /home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEM.hpp:214:0,
                 from /home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Cell/Intrepid2_CellTools.hpp:67,
                 from /home/rppawlo/Trilinos/packages/panzer/dof-mgr/src/Panzer_NodalFieldPattern.cpp:45:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp: In constructor ‘Intrepid2::Basis_HGRAD_QUAD_C1_FEM<ExecSpaceType, outputValueType, pointValueType>::Basis_HGRAD_QUAD_C1_FEM()’:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:181:7: error: ‘ordinal_type_array_1d_host’ was not declared in this scope
       ordinal_type_array_1d_host tagView(&tags[0], 16);
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:184:7: error: ‘ordinal_type_array_2d_host’ was not declared in this scope
       ordinal_type_array_2d_host ordinalToTag;
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:185:7: error: ‘ordinal_type_array_3d_host’ was not declared in this scope
       ordinal_type_array_3d_host tagToOrdinal;
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:186:31: error: ‘tagToOrdinal’ was not declared in this scope
       this->setOrdinalTagData(tagToOrdinal,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:187:31: error: ‘ordinalToTag’ was not declared in this scope
                               ordinalToTag,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:188:31: error: ‘tagView’ was not declared in this scope
                               tagView,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:204:17: error: ‘basisCardinality_’ was not declared in this scope
       dofCoords(basisCardinality_,basisCellTopology_.getDimension());
                 ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_QUAD_C1_FEMDef.hpp:204:35: error: ‘basisCellTopology_’ was not declared in this scope
       dofCoords(basisCardinality_,basisCellTopology_.getDimension());
                                   ^
In file included from /home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEM.hpp:223:0,
                 from /home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Cell/Intrepid2_CellTools.hpp:68,
                 from /home/rppawlo/Trilinos/packages/panzer/dof-mgr/src/Panzer_NodalFieldPattern.cpp:45:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp: In constructor ‘Intrepid2::Basis_HGRAD_HEX_C1_FEM<ExecSpaceType, outputValueType, pointValueType>::Basis_HGRAD_HEX_C1_FEM()’:
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:239:7: error: ‘ordinal_type_array_1d_host’ was not declared in this scope
       ordinal_type_array_1d_host tagView(&tags[0], 32);
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:242:7: error: ‘ordinal_type_array_2d_host’ was not declared in this scope
       ordinal_type_array_2d_host ordinalToTag;
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:243:7: error: ‘ordinal_type_array_3d_host’ was not declared in this scope
       ordinal_type_array_3d_host tagToOrdinal;
       ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:244:31: error: ‘tagToOrdinal’ was not declared in this scope
       this->setOrdinalTagData(tagToOrdinal,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:245:31: error: ‘ordinalToTag’ was not declared in this scope
                               ordinalToTag,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:246:31: error: ‘tagView’ was not declared in this scope
                               tagView,
                               ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:262:17: error: ‘basisCardinality_’ was not declared in this scope
       dofCoords(basisCardinality_,basisCellTopology_.getDimension());
                 ^
/home/rppawlo/Trilinos/packages/intrepid2/refactor/src/Discretization/Basis/Intrepid2_HGRAD_HEX_C1_FEMDef.hpp:262:35: error: ‘basisCellTopology_’ was not declared in this scope
       dofCoords(basisCardinality_,basisCellTopology_.getDimension());
                                   ^
make[2]: *** [packages/panzer/dof-mgr/src/CMakeFiles/panzer-dof-mgr.dir/Panzer_NodalFieldPattern.cpp.o] Error 1
make[1]: *** [packages/panzer/dof-mgr/src/CMakeFiles/panzer-dof-mgr.dir/all] Error 2
make: *** [all] Error 2

```

**Comment by kyungjoo-kim** [comment]
Oh.... yesterday I just started working on tri and tet. While doing that, I also started converting the code to uvm-free. I presumed that the Tag data is used in host space and we need the new kokkos mirror view and I am waiting for the update on develop branch. It will be fixed in this morning. Sorry I do not expect that you directly compile the code but thought that you first change field container with md field wrapper.

**Comment by rppawlo** [comment]
OK. I will test with the refactored intrepid disabled for now. I am just changing the field containers directly to DynRankViews (instead of mdfields) but figured I could get you info on the refactor at the same time. Just to make sure I understand - to turn on the refactored intrepid2 branch, I need to wait until there is a sync from kokkos into trilinos later today?

**Comment by kyungjoo-kim** [comment]
Hi Roger, I fixed all the broken things... 

You need a temporary patch and this patch is working with GCC. I emailed you the patch. Nathan told me that an intel compiler does not like his fix. Now triangle and tets are enalbed for basis and integration.

**Comment by rppawlo** [comment]
Thanks @kyungjoo-kim I reverted to using intrepid2/core for now. The replacement script was just the beginning. Since Views require strings in the ctor, all the temporary FieldContainers now need a name. Not to mention intrepid factories and panzer factories now require more template parameters (intrepid defaulted to FieldContainers). might take a little bit until I can switch back to refactor branch. Hopefully today or tomorrow.

**Comment by mperego** [comment]
Thanks @rppawlo. I will also need to do this for Albany. So, I'm planning to exploit your experience with this :) !

**Comment by kyungjoo-kim** [comment]
I just pushed all changes for tri and tet in cell tools and function space tools. 
- hgrad tri and tet are done
- cubature tri and tet (default cubatures) are done
- celltools now have tri and tet
- functionspacetools now have tri and tet

@ndellingwood will fix the constructor ambiguity and push to develop branch later today. Until then, the patch I sent should be okay and I tested against with the patch. 1st order Hdiv and Hcurl for tri and tet are on the way....now.... maybe.... probably....

After hdiv and hcurl on tri and tet, panzer can proceed tests.

**Comment by hcedwar** [comment]
FYI: Kokkos::View can be constructed without string labels if needed.

**Comment by rppawlo** [comment]
@hcedwar What about DynRankView? Could you add that support too?

**Comment by ndellingwood** [comment]
@rppawlo I'll work on that support for the DynRankView

Resolved the issues with constructor ambiguity compilers had related to create_mirror_view functionality added to DynRankView, but another issue popped up in unit testing I am working on now.

**Comment by rppawlo** [comment]
Great! Thanks @ndellingwood ! Not to pile it on, but one other painful part of switching from Intrepid::FieldContainers to DynRankView is the lack of a member function called resize - although our use of resize is more like the realloc function. Don't copy the data but change the extents. So for now where we have 
```
a.resize(dim0,dim1);
```
I have been replacing with 
```
realloc(a,dim0,dim1);
```
So far so good.  But the transition would have been easier with this method on the View and DynRankView. Any chance that something like this could be added as well? Maybe I missed something?

**Comment by crtrott** [comment]
Carter, I think what Roger would want is a constructor which only takes dimension arguments. I am against adding that, because if we give people a way to do some common task with less effort they will take it. For now that might not be bad, but if we find ourselves in two years with a nasty memory issue deep in a million line code, everybody will have created views without labels, and suddenly it becomes much harder to track down where it is going wrong. So I would rather force developers for their own good to give names.

**Comment by ndellingwood** [comment]
@rppawlo an alternative that may be helpful to avoid manually adding labels but to still have labels could be to use a macro similar to what @kyungjoo-kim used in the refactored intrepid2 unit tests. 
For example from ArrayTools test_01:
`#define ConstructWithLabel(obj, ...) obj(#obj, __VA_ARGS__)
typedef Kokkos::DynRankView<value_type,DeviceSpaceType> DynRankView;
 DynRankView ConstructWithLabel(a_2_2, 2, 2);
// this will create DynRankView("a_2_2", 2, 2);`

Would this be satisfactory alternative?

**Comment by ndellingwood** [comment]
Reposting the code snippet with line breaks, not sure why they didn't come through...
`#define ConstructWithLabel(obj, ...) obj(#obj, __VA_ARGS__)`
`typedef Kokkos::DynRankView<value_type,DeviceSpaceType> DynRankView;`
`DynRankView ConstructWithLabel(a_2_2, 2, 2); // this will create DynRankView("a_2_2", 2, 2);`

**Comment by mperego** [comment]
@rppawlo, just a warning about resize, not sure it is relevant for Panzer. In a few places in Intrepid tests, resize was used as Matlab reshape.. to go from, say, a three-dimensional array of dimension (1, dim1, dim2) to a two-dimensional view of dimension (dim1, dim2), but retaining the data content. In this cases you can use subviews.

**Comment by rppawlo** [comment]
@ndellingwood - I could do that but it doesn't get around the real problem. I'm trying to convert a large code base with a minimum amount of manual labor. We have many functions that create temporary arrays. The all used a common constructor without a string (or a default ctor and then called resize) so that we could template on the array and it would work for different array types. I used a sed script to convert from FieldContainer to DynRankViews, but I still need to manually change the temporary array ctors to supply a string. If I use the macro above, I suspect that we would still need to manually change the constructor.   

```
   FieldContainer ublock0, tblock0, tblock1;
   ublock0.resize(localIds["eblock-0_0"]->size(),4);
   tblock0.resize(localIds["eblock-0_0"]->size(),4);
   tblock1.resize(localIds["eblock-1_0"]->size(),4);
```

**Comment by kyungjoo-kim** [comment]
@rppawlo I also thought a similar thing about "resize" and at the end I just explicitly create and reassign the view. Probably a simple function can do the job too, something like
```
template<typename ViewType>
void realloc(ViewType a, int dim0, int dim1) {
   a = ViewType(a.label(), dim0, dim1);
} // and create this for all dimension cases
```
or 
```
#define ReallocView(obj, ...) obj = typename decltype(obj)(obj.label(), ...);  
```
This means that the obj has label when it is created through the default constructor. I did not test but it probably works.

**Comment by crtrott** [comment]
Doesn't that throw away the existing data?

**Comment by rppawlo** [comment]
@mperego - we are using them in functions that take an incoming array that may not have the correct extents. So to be safe we always resize the array. 

@kyungjoo-kim - yes - that is exactly what I have been doing. Kokkos View and DynRankView both support a non-member realloc function. Thanks to Koenig lookup, we can even use it in code templated on the array type!

**Comment by kyungjoo-kim** [comment]
Oh... yeah it throws data (so the name is wrong "AllocView"). I found that the resize function in Intrepid is mostly used for dynamic allocation. Also, the resize function also do reshaping arrays, which is not really right in the view context. If the resize perform reshaping with preserving data, that cannot be fixed by "sed" magic. Since roger got a pain from temporary containers (one time use mostly), "resize" actually means "allocate".

**Comment by kyungjoo-kim** [comment]
Ah. Btw i jusr realize the case that resize actually does resize. Often intrepid requires exact cardinality and dimensions. To match the dimensions the panzer may do resize. For this case subview can do the job.

**Comment by hcedwar** [comment]
For clarity:  Is the need for a _reallocate_ or _reconstruct_ function that simply uses the existing memory if sufficiently large and replaces the multidimensional array referenced by the View?  This means that the contents of the old array are no longer needed and can be destroyed (but not deallocated).
If this is the case then should the old contents be re-initialized for the new array?

**Comment by rppawlo** [comment]
@hcedwar - yes. We are writing functions that are going to fill an incoming view. However the user may not always allocate the size correctly if at all. So to be safe, we have just called the resize() function on the intrepid FieldContainer to make sure it had the right dimensions and could hold all values. We don't care what the original values were and we don't need to reallocate if the capacity is sufficient. We are overwriting the values. These are usually for small arrays during a setup phase (that is in the noise of the overall calculation) so we  care more about robustness about speed.

**Comment by ndellingwood** [comment]
@rppawlo @kyungjoo-kim @mperego 
The update to DynRankView for create_mirror and create_mirror_view of host on device (needed for refactored intrepid2) is now in the Kokkos develop branch.

**Comment by rppawlo** [comment]
Panzer is done. One of the applications are done. One more application to go and I can push.

**Comment by kyungjoo-kim** [comment]
@ndellingwood Could you give us the update on the Intrepid2 basis functions (i.e., Hcurl for the wedge element and 2nd order hgrad functions) so that Roger can figure out which one can be tested (or commented out from Panzer for now) ?

**Comment by ndellingwood** [comment]
@rppawlo @kyungjoo-kim 
I spent the past couple days working on a kokkos feature, will dedicate next few days to the basis functions, should finish working on the Hcurl wedge tomorrow and will start 2nd order hgrad when that is complete.

**Comment by ndellingwood** [comment]
@rppawlo @kyungjoo-kim 
HCURL_WEDGE_I1 basis and unit test finished and pushed to Trilinos master, pass with Cuda. 2nd order HGRAD in-progess.

**Comment by ndellingwood** [comment]
@rppawlo @kyungjoo-kim 
We finished the C2 elements. 
All intrepid2 basis and unit tests do not rely on UVM when using Cuda.

**Comment by rppawlo** [comment]
Just pushed the changes to panzer and all applications that depend on panzer. The Intrepid2::FieldContainer is now completely removed. Stage 1 of transition is complete. Stage 2 will be the switch from Intrepid2/core to Inrepid2/refactor. Since this requires a special configure flag, we will have to coordinate with the applications and all automated testing to add this flag or change the default to point to the refactor branch.

**Comment by rppawlo** [comment]
here is the updated script for refactoring


[refactor2.sh.txt](https://github.com/trilinos/Trilinos/files/271736/refactor2.sh.txt)

**Comment by rppawlo** [comment]
@kyungjoo-kim and @mperego - I started on stage 2 of the refactor. In building the Intrepid2 code in the refactor directory, I got the following failures for an openmp build:

```
96% tests passed, 4 tests failed out of 92

Label Time Summary:
Intrepid2    =  19.47 sec (92 tests)

Total Test time (real) =  19.51 sec

The following tests FAILED:
	  5 - Intrepid2_refactor_unit-test_Discretization_FunctionSpaceTools_Serial_Test_04_MPI_1 (Failed)
	 10 - Intrepid2_refactor_unit-test_Discretization_FunctionSpaceTools_OpenMP_Test_04_MPI_1 (Failed)
	 62 - Intrepid2_refactor_unit-test_Discretization_Integration_Serial_Test_06_MPI_1 (Failed)
	 69 - Intrepid2_refactor_unit-test_Discretization_Integration_OpenMP_Test_06_MPI_1 (Failed)
Errors while running CTest
```
Is this expected?

**Comment by ndellingwood** [comment]
@rppawlo @kyungjoo-kim @mperego 
I compiled and ran the tests (OpenMP, Intel compiler) and had the same failures. 
I checked the output of FunctionSpaceTools_OpenMP_Test_04_MPI_1 and it looks like the error tolerance could be the issue, is this due to the changes in #397 ? 

The changes made to Intrepid2 in that commit were for reported compile errors in Albany for multiple definitions of Intrepid2::Parameters::Epsilon, Threshold, and Tolerance.
Code change snippet:
```
 -    static const double Epsilon;
-    static const double Threshold;
-    static const double Tolerence;
+    static constexpr double Epsilon   = 1.0e-16; 
+    static constexpr double Threshold = 1.0e-15;
+    static constexpr double Tolerence = 1.0e-14;
```


Posted below is some sample error output:
FunctionSpaceTools_OpenMP_Test_04_MPI_1
```
Volume of cell 0 =   0.16666666666666665741    vs.    Analytic value =    0.16666666666666665741
Volume of cell 1 =    64.666666666666657193    vs.    Analytic value =     64.666666666666671404
Volume of cell 2 =  0.066666666666667012686    vs.    Analytic value =   0.066666666666666665741
```

Intrepid2_refactor_unit-test_Discretization_Integration_OpenMP_Test_06_MPI_1 
```
Cubature order 20 integrating x^1  * y^3  * z^9 :   1.0406260406149585e-07   1.0406260406260406e-07   1.1082e-18   <?   1.0406e-18
                                                                                                      ^^^^---FAILURE!
Cubature order 20 integrating x^1  * y^3  * z^10:   6.1213296506775731e-08   6.1213296507414156e-08   6.3842e-19   <?   6.1213e-19
                                                                                                      ^^^^---FAILURE!
Cubature order 20 integrating x^1  * y^3  * z^11:   3.7408125643000799e-08   3.7408125643419763e-08   4.1896e-19   <?   3.7408e-19
                                                                                                      ^^^^---FAILURE!
```

**Comment by ndellingwood** [comment]
@rppawlo @kyungjoo-kim @mperego 
For the unit tests from the mentioned Intrepid2 files, the error tolerance is set to Parameters::Tolerance

`const auto tol = Parameters::Tolerence;`

Changing the hardcoded values for Epsilon, Threshold, and Tolerance from #397  commit 17d0039 may affect Albany tests, so a solution would be to adjust the tolerance in the Intrepid2 unit tests, but I will wait to hear back before making any changes.

**Comment by ndellingwood** [comment]
Test failures now with Cuda also:

```
The following tests FAILED:
	  5 - Intrepid2_refactor_unit-test_Discretization_FunctionSpaceTools_Serial_Test_04_MPI_1 (Failed)
	 10 - Intrepid2_refactor_unit-test_Discretization_FunctionSpaceTools_Cuda_Test_04_MPI_1 (Failed)
	 61 - Intrepid2_refactor_unit-test_Discretization_Integration_Serial_Test_06_MPI_1 (Failed)
	 68 - Intrepid2_refactor_unit-test_Discretization_Integration_Cuda_Test_06_MPI_1 (Failed)
	 81 - Intrepid2_refactor_unit-test_Shared_RealSpaceTools_Cuda_Test_01_MPI_1 (Failed)
Errors while running CTest
```

Intrepid2_refactor_unit-test_Shared_RealSpaceTools_Cuda_Test_01_MPI_1 
 is a new failure (not matching OpenMP/Serial), its output shows failure for math correctness in test 2, 
```
===============================================================================
| TEST 2: correctness of math operations                                      |
===============================================================================
```
```
...
-- Checking inverse, subtract, and vectorNorm 

INCORRECT inverse OR subtract OR vectorNorm
...
```




Other tests have similar error output to matching OpenMP failures, for example
Intrepid2_refactor_unit-test_Discretization_FunctionSpaceTools_Serial_Test_04_MPI_1
```
===============================================================================
| TEST 1: correctness of cell volumes                                         |
===============================================================================
Volume of cell 0 =   0.16666666666666665741    vs.    Analytic value =    0.16666666666666665741
Volume of cell 1 =    64.666666666666657193    vs.    Analytic value =     64.666666666666671404
Volume of cell 2 =  0.066666666666666540841    vs.    Analytic value =   0.066666666666666665741

End Result: TEST FAILED
```

**Comment by agsalin** [comment]
I hardcoded these values as a workaround to get the code to compile. The previous code, of setting a static const and then assigning a const value to it by calling the epsilon function, did not compile for me. The error I ran into was when the header was included multiple times, it complained of multiple definitions of the same variable. I suspect that the unit tests don't test for this. I wasn't able to find a way to set a static const by calling the epsilon() function, as was the case before, so hardcoded to what I thought would be similar values. Maybe someone can restore the old values by some magic.

Sorry for suggesting the fix in a pull request without running the unit tests.

**Comment by kyungjoo-kim** [comment]
It looks like that those test failures are due to different use of threshold from previous one. Let me think about this in the coming next week. Since the code produce the right results (machine precision set up is only difference), it will not block your transition phase 2.

**Comment by kyungjoo-kim** [comment]
I changed unit tests use a function to get tolerence remaining the hard-wired Tolerence. This seems to be the compiler issue with C++ standards. 

Nathan and Roger

Could you check the unit tests again with my update today ? 

Andy,

The hardwired Tolerence is still there and you can still compile your code. Eventually we need to fix this problem. BTW, can you compile Intrepid2 unit tests ? Just wondering.

**Comment by ndellingwood** [comment]
@kyungjoo-kim Unit tests passing with Intel 16.0.1 and Cuda 7.5.18

**Comment by rppawlo** [comment]
All tests passed for me with gcc 5.2 openmp build. Thanks @kyungjoo-kim !

**Comment by kyungjoo-kim** [comment]
@rppawlo @ndellingwood @agsalin 

Today I talked with Andy and found a few issues regarding to dynrankview, mdfield, and intrepid2.

1. MD field has two version of wrapping kokkos view.
- MDField<datatype>  wraps dynrankview
- MDFIeld<datatype,dim1,...> wraps kokkos view and provide operator for different rank objects

These two essentially do the same job. Roger, are you going to deprecate latter one that does not use dynrankview ? or both of them are used together ?

One advantage of using MDField<datatype,dim1,dim2...> is on readability of code as the declaration of the container dictates the intrepid array structure (CNP, CNPD, CNPDD, etc.). A disadvantage of using two versions of MDField is on inter-operatibility between view and dynrankview (as intrepid2 explicitly requires dynrankview, there needs conversion between them -- not an easy problem).

2. Possible solution is to match the internal kokkos view in MDField<datatype,dim1,dim2...> to dynrankview. The templated dimensions are used for constructor for MDField. Do you have any other reason that we need to maintain the version of static kokkos view ?

**Comment by rppawlo** [comment]
There are essentially three versions of MDField:
1. The static mdfield that wraps a kokkos View
2. The deprecated runtime mdfield that wraps kokkos View for each possible rank
3. The new runtime mdfield that wraps the kokkos DynRankView.
 
Versions 2 and 3 can NOT exist at the same time. An ifdef will enable one or the other. Version 1 is always available. Version 1 should be completely interoperable with version 2 or 3 based on how we allocate the fields. See 

https://github.com/trilinos/Trilinos/blob/master/packages/phalanx/src/Phalanx_MDField_DynRank_Def.hpp#L361

Christian and I went over this code and had to fix last week due to the changes to DynRankView in the latest Kokkos snapshot. Version 3 is intended to be used with the Intrepid2 refactor branch. I left version 2 in the code but disabled by default in case the transition to DynRankView broke any codes (in particular for Albany since I tested all the other codes that use Phalanx). My plan was to remove version 2 in the future, however if DynRankView in Kokkos goes away I may now keep it around as a fallback (a completely interoperable fallback since 1 and 2 both use static Views underneath.

I want to keep the static version 1 around for speed. As the Kokkos team points out, the DynRankView is slower. So I do want to keep everything as fast as possible when it is possible - which is the bulk of our assembly code that does not use Intrepid2.

Where are you seeing interoperability issues? Maybe we can meet tomorrow to discuss?

**Comment by kyungjoo-kim** [comment]
One major problem is 

```
MDField<datatype> drank_view(C,P,D);
MDField<datatype,C,P,D> srank_view;

// let's say copy
Intrepid2::copy( drank_view.get_kokkos_view(), srank_view.get_kokkos_view() );

// this implies
Intrepid2::copy( DynRankView, View );

// however all intrepid functions are declared as like 
template<ValueTypeA, ...PropertiesA, ValueTypeB, ...PropertiesB>
Intrepid2::copy( DynRankView<ValueTypeA, ...PropertiesA>, DynRankView<ValueTypeB, ...PropertiesB> )

// dynrankview types are deduced from input. if we use static kokkos view, it cannot be deduced.
// this means i need to provide interface layer. this is not a good option. 
// I enforce dynrankview as we all want to use the same container to call intrepid layer. 
// if I assume plain kokkos view can be an input of intrepid2, there will be some problems
template<ViewTypeA, ViewTypeB>
Intrepid2::copy( ViewTypeA A, ViewTypeB B) {
  // here copy constructor is required to change arbitray view to dynrankview. 
  Intrepid2::copy( DynRankView<ViewTypeA::value_type, ViewTypeA::Properties...>(A), 
                            DynRankView<ViewTypeB::value_type, ViewTypeB::Properties...>(B) );
}

// or users do appropriately convert view to dynrankview to call intrepid2 functions.
// the main issue is we need to modify this change with bare eyes and it is quite 
// annoying to app developers as they need to find how the object is declared like this..
Intrepid2::copy( A, DynRankView<MDField::array_type::value_type, MDField::array_type::properties...>(B)); 
```

I am not sure if I write my concerns well.

**Comment by rppawlo** [comment]
That expliains it perfectly. We have a few options:
1. Use only Dynamic MDFields - probably lots of work to switch over
2. I can add a method to the static MDField called get_kokkos_dyn_rank_view() that will convert the statically allocated View into a DynRank view. It would use the exact same code as I linked to i nmy last comment. All dynamic mdfields already use statically allocated view anyway.
3. I could have two implementations of the static MDField and the application could choose at compile time whether the underlying kokkos object is a View or DynRankView. 

I think option 2 is painless and very simple to implement. I hate option 3 as it adds more complexity, requires multiple builds to test against due to more ifdefs and requires more code to maintain. Option 1 could just be done now if you want to just use all DynRankViews. Any opinion? I can add option 2 tomorrow - it will be very easy to add and test.

**Comment by kyungjoo-kim** [comment]
I summarize yesterday discussion (Andy, Eric C. Nathon and me) briefly. 

1. To resolve this inconenience, the best option is some implicit conversion from MDField to DynRankView. I am not sure if this is techniacally possible or not (the conversion is not intuitive as it involves template deduction and conversion of type together). 

However, if this works, we won't need to put a.get_kokkos_view. This also prevent cumbersome distinction of two objects when users use both MDField or DynRankView directly for temporary variable.

2. If implicit conversion does not work, we still need to extract get_kokkos_view. In this case, a question is what is the right default behavior. Eric suggest that get_kokkos_view should return dynrankview. If one want to use static view for performance reason, that might be the special case and suggest to use get_kokkos_static_view. Roger suggests the other way get_kokkos_dyn_rank_view. Since this is Panzer and application level modification, you can vote. Anyway if the first option is possible, we don't even need to use get_kokkos_view.

3. Andy also points out createDynRankView (or createView) uses get_kokkos_view. Here again there is ambiguity as get_kokkos_view can return either static view or dynrankview. Since the instance of kokkos view is required to extract meta data (e.g., AD type and dimensions), he asked if there is a more elegant solution to extract such properties.

4. Jacobian sign flip. In the refatored version I return error when it see negative jacobians. Andy said there are two reasonable cases to use negative jacobian. The first one is periodic boundary condition in Albany. The elements on the periodic boundary are flipped according to use the nodes on the other side. The other case is non-linear elasticity. If an object is largely deformed during nonlinear iteration (which produces negative jacobian), the code need to stop the iteration and reduce the step size to proceed the solver. Clearly invoking an error for negative jacobian is not a good idea. However, we still need to check negative jacobians when they are indeed an error.

The solution might be if you don't need to know which cell has a negative jacobian, I can simply return a flag for this. Is this okay ? 

5. We want to evaluate overhead in using dynrankview and compare it against kokkos view. The overhead might come from using additional zero indices and subview operations e.g., ( n1, n2) vs ( n1, n2, 0, 0, 0, 0 ,0 ,0, 0). This information might be useful for future decision making.

**Comment by rppawlo** [comment]
I attempted the implicit conversion. The template argument deduction was failing. If I removed the templates, then I could get this to work, so long as the arguments were not non-const references. If they were non-const references, then I would get a compile failure due to binding rvalue to non-const reference. Will talk with Eric P to see if he has any suggestions on how to work around these issues.

**Comment by eric-c-cyr** [comment]
@rppawlo Did you just try this today, or is it something you had tried a while back? I was hoping this was possible, but wasn't entirely sure. In particular, its not possible to introduce a copy constructor for DynRankView.

**Comment by rppawlo** [comment]
Just tried it today and haven't explored much at this point. If I take the templates out, I can get the following to work:
```
MDField<double> mdf_a("df_a", node_scalar);
MDField<double> mdf_b("df_b", node_scalar);      mdf_a.setFieldData(PHX::KokkosViewFactory<double,PHX::Device>::buildView(mdf_a.fieldTag()));
mdf_b.setFieldData(PHX::KokkosViewFactory<double,PHX::Device>::buildView(mdf_b.fieldTag()));
Kokkos::DynRankView<double,PHX::Device> drv_a = mdf_a;
Kokkos::DynRankView<double,PHX::Device> drv_b = mdf_b;
addDRV(drv_a,drv_b);
addDRV(mdf_a,mdf_b);
```
where addDRV is hard coded to take DynRankView objects. If I template the scalar type, we have issues with type deduction. We may be able to get this to work. I just haven't spent much time on it yet.

**Comment by rppawlo** [comment]
Just tried it today and haven't explored much at this point. If I take the templates out, I can get the following to work:
```
MDField<double> mdf_a("df_a", node_scalar);
MDField<double> mdf_b("df_b", node_scalar);
mdf_a.setFieldData(PHX::KokkosViewFactory<double,PHX::Device>::buildView(mdf_a.fieldTag()));
mdf_b.setFieldData(PHX::KokkosViewFactory<double,PHX::Device>::buildView(mdf_b.fieldTag()));
Kokkos::DynRankView<double,PHX::Device> drv_a = mdf_a;
Kokkos::DynRankView<double,PHX::Device> drv_b = mdf_b;
addDRV(drv_a,drv_b);
addDRV(mdf_a,mdf_b);
```
where addDRV is hard coded to take non-ref DynRankView objects. If I template the scalar type, we have issues with type deduction. We may be able to get this to work. I just haven't spent much time on it yet.

**Comment by kyungjoo-kim** [comment]
I try to resolve the item 4 and 5. Now functionspacetools::computeCellMeasure will return a bool flag and all cell measure will be positive. If it found negative jacobian, the routine return "true" (hasNegativeJacobian), otherwise it returns "false". So, if you want to check negative jacobian, wrap the routine in the assert.

Regarding to the performance comparison on view and dynrankview, there is a sad news. dynrankview is almost 5 times slower than view. I compute determinant in the same way I compute in Intrepid2 with using view and dynrankview. I tested both version : one is using subdynrankview and the other is using A(C, P, i, j) to compute determinant. Both of them shows similar performance. So the main overhead is from the operator() as suspected. I put this test in perf-test in Intrepid2/refactor. You can compile the test with cmake example flag is on. 
```
Test Array Structure (C,P,D) = 4096, 9, 3
===================================================
TEST 0: RealSpaceTools det
        input (C,P,D,D), output (C,P)
===============================================================================
TEST 0:  t_view = 0.045074,  t_dynrankview = 0.255936, ratio (v/drv) = 0.176114
===============================================================================
```

**Comment by kyungjoo-kim** [comment]
I just take another test that Andy suggest to find out where the performance problem is.

This time I uses rank 8 view to confirm the overhead comes from operator(). I can see the result is similar. So adding extra zeros in operator() is not good. @ndellingwood @hcedwar Do we have anyway to avoid them ? 

```
inViewType in("inView", C,P,D,D,1,1,1,1);
outViewType out("outView", C,P,1,1,1,1,1,1);

// and change the determinant code access like 
// Mat(C,P,i,j,0,0,0,0);

===============================================================================
TEST 0:  t_view = 0.228621,  t_dynrankview = 0.198439, ratio (v/drv) = 1.1521
===============================================================================
```

**Comment by hcedwar** [comment]
This is the inherent performance problem with dynamic rank view.
This is why it is recommended to use View when performance is required.

**Comment by kyungjoo-kim** [comment]
But we can fix this problem directly call the internal implementation in kokkos, isnt' it ?

**Comment by hcedwar** [comment]
That is already happening - it is calling the implementation for a rank seven view and the optimizer is obviously not eliminting the hard-wired 0 arguments.  This is not surprising.

**Comment by crtrott** [comment]
Which compilers did you use? Can you try Intel 16 and make sure to use -O3?

**Comment by kyungjoo-kim** [comment]
Previous run was on my local workstation with "-g -O3". It is also possible that the "-g" option still leave symbols for added 0 arguments. Computation part uses subview, which means all views have stride layout. Is this going to affect the performance even if we provide zero arguments ?

I tested with Intel 16 with "-O3" only on Compton.  The time ratio does not change. BTW, intel 16 is so slow in configuration.

**Comment by nmhamster** [comment]
Intel 16 slowness on Compton is due to license grabbing. We are looking into how to address that issue.

Have you tried this test on Intel 17.0 on Shepard as well? This contains better optimization.

It is unlikely that -g fundamentally changes optimization behavior with respect to 0 elimination. Intel have worked very hard to ensure the optimizations are still applied at the cost of very large debug table generation (which is why we have rediculous binary sizes).

**Comment by crtrott** [comment]
Oh strided Layout can have a dramatic impact on performance. But if you use strided in either case (with normal views or dynrankview) than that is not the issue. It rather points to the compiler not getting rid of the zero arguments because the expression is to complicated or so (we are talking about a 8 dim index calculation after all). Unfortunate but not totally unexpected as Carter said. That is the problem with buying convenience through very generic code constructs.

**Comment by kyungjoo-kim** [comment]
I commonly use subviews (or subdynrankviews) to extract rank 2 matrix from rank 4 a collection of matrices. The strided layout is not a real issue (anyway the memory is still contiguous). 

I agree that the performance problem comes from the extra zeros. Then, a couple of questions are...
1. Do we want to stick to dynrankview ? At this moment, it is quite difficult to support both dynrankview and static views. The subview syntax is different and the behavior is also different (subdynrankview allows additional range arguments but static view does not allow it). Rank detection mechanism is also different and quite amout of refactored code would need additional changes.

2. Can we expect performance improvement in operator() ? I thought that this is an easy problem if kokkos view allows private variable access from dynrankview so that it compute indices directly (e.g., m_map) without using base rank 8 operator().

3. Last questions. From these tests, I can roughly expect at least 3 timres slowness compared to static views. Probably the overhead increases more for more complicated computations (a larger number of accesses to operator()). But the code is much cleaner than before and it will work on gpus. Is this acceptable ?

**Comment by rppawlo** [comment]
So there is also the alternate design that we used for dynamic rank views in phalanx. Allocate each supported rank as a static view in the underlying implementation of DynRankView. As @crtrott mentioned, this will use more memory and add to register pressure, but it will eliminate the extra zeroes. We can easily compare as I left the old implementation in phalanx. You can compile the static implementation or the new DynRankView implementation via a configure option:

 -D Phalanx_ENABLE_KOKKOS_DYN_RANK_VIEW:BOOL=OFF

Default is to use the new DynRankView underneath.

I would be happy to eliminate the DynRankView long term, but this is probably an important capability for code transition to kokkos.

**Comment by etphipp** [comment]
My suggestion would be to gradually refactor Intrepid2 to use the static-rank view, and only use the dynamic rank view in the interim.  With improved interoperability between the two, I think that would be feasible, and in the long run give the best performance.  For this to work efficiently from a developer perspective most of the stuff I mentioned in issue [#296](https://github.com/kokkos/kokkos/issues/295), is needed, in particular:

1. The ability to assign a static-rank view to a dynamic-rank view
2. The ability to deep_copy between the two
3. A unified way of computing the rank
4. A unified way of doing subviews (through SFINAE it should be possible to overload subview() for dynamic-rank views instead of the new subdynrankview() function, although it probably requires a very small change to the declaration of subview()).

**Comment by ndellingwood** [comment]
I'm working on the interoperability requests this week, will let you know as they're ready.

**Comment by kyungjoo-kim** [comment]
I tested again with some modification in kokkos
```
/// instead of using this
return view_type::operator()(i0,i1,0,0,0,0,0); 
/// using this
return implementation_map().reference(i0,i1);
```

I got 
```
 Performance Test for measuring DynRankView Overhead
 # of workset = 8
 Test Array Structure (C,P,D) = 4096, 9, 3
===============================================================================

===============================================================================
TEST 1: RealSpaceTools det
        input (C,P,D,D), output (C,P)
===============================================================================
 -> Testing View
 -> Testing DynRankView
===============================================================================
TEST 0:  t_view = 0.043913,  t_dynrankview = 0.043238, ratio (v/drv) = 1.01561
===============================================================================
End Result: TEST PASSED
```

There is no performance disadvantage. I want you double check the usage of reference is okay for all layouts and ad types. Then, dynrankview can survive.

**Comment by crtrott** [comment]
I believe that change should be fine.

**Comment by kyungjoo-kim** [comment]
Now we know that dynrankview does not have performance disadvantages as we removed the extra zeros in operator(). Let's recall the first discussion about mdfield and kokkos view implicit conversion. 

Based on the Roger's trial and some personal googling, the implicit conversion with template arguments is quite difficult to make it work. So, I think that panzer needs to clarify what type is returned and what syntax is supposed e.g., "get_kokkos_view" or "get_kokkos_dynrankview" or "get_kokkos_static_view". I presume that panzer still wants to use its own dynamic rank implementation (do we keep that version ?). @rppawlo @eric-c-cyr Could you comment about this issue ?

**Comment by eric-c-cyr** [comment]
I prefer "get_kokkos_view" and "get_kokkos_static_view". The first one is the 99% use case for those calling Intrepid. It should be as simple as possible. I'm not sure about the dynamic rank implementation that you mentioned. Perhaps @rppawlo can comment on that.

**Comment by rppawlo** [comment]
I agree with @eric-c-cyr . Should we trim the call even more?
get_view()
get_static_view()

I can put this in today and test against all the codes.

**Comment by kyungjoo-kim** [comment]
Okay. Then, I can now assume that get_kokkos_view() or get_view() returns dynrankview. 

md field dynrank implementation uses the MDField<datatype,dim0,dim1> construtor while MDField<datatype> uses kokkos dynrankview internally. 
  
For panzer and phalax tests, kokkos needs to provide the copy constructor. @ndellingwood Do we have time line for this (when this is updated in kokkos develop branch and when this is merged to trilinos) ?

After these things are done, @agsalin can proceed Albany conversion work.

**Comment by ndellingwood** [comment]
@kyungjoo-kim Working on the copy constructor, will ask Christian when he anticipates testing for the next snapshot of Kokkos into Trilinos.

**Comment by kyungjoo-kim** [comment]
I thought that mdfiield requires kokkos support for the copy constructor (some proper reference counting and padding as discussed in tpetra meeting) but it seems like that @rppawlo can do this independently by extracting the raw pointer and wrapping it with dynrankview. Thanks.

**Comment by rppawlo** [comment]
We have been using this technique successfully for quite a while. Behind the scenes we have been allocating all views statically and converting to DynRankView as needed by pulling out the pointer. When padding is reenabled, this will break as we discussed in kokkos meeting today and we will need the assignment operator to switch views from static to dynamic. @ndellingwood is working on this now I believe.

**Comment by rppawlo** [comment]
Just pushed the changes to mdfield. 

get_view() always returns dynamic. 
get_static_view() always returns static.

Since there is no performance penalty, I will be pulling out the deprecated phalanx implementation of dynamic rank mdfield. The dynamic rank mdfield using the underlying kokkos DynRankView will obviously remain.

**Comment by ndellingwood** [comment]
Update on DynRankView interoperability with Views [](https://github.com/kokkos/kokkos/issues/295)

Working on the some of the interoperability feature requests (subdynrankview to subview, copy and assignment of View to DynRankView) started to require multiple design workarounds, including some unacceptable within the Kokkos View (e.g. exposing the shared allocation tracker). After talking with Carter, the best route forward will be to redesign the DynRankView so that it no longer privately inherits the rank 7 View but will essentially be a new form of View (rank 7) that shares most all of its internal implementation with the original View.

**Comment by agsalin** [comment]
@rppawlo: It looks like you forgot to to change Phalanx_MDField_DynRank.hpp to also use get_view() instead of get_kokkos_view(). Or maybe I am not following the discussion.

**Comment by rppawlo** [comment]
are you on the develop branch?

**Comment by agsalin** [comment]
Both branches are the same: the dynamic MDField definition in Phalanx_MDField_DynRank.hpp has a method get_kokkos_view() and not get_view().

**Comment by kyungjoo-kim** [comment]
Try 'git show master' and if there is difference.

**Comment by rppawlo** [comment]
Darn. I did miss that for dynamic. Will push fix later today as part of a larger change set. Sorry about that.

**Comment by rppawlo** [comment]
@agsalin fix is pushed.

**Comment by ndellingwood** [comment]
Finished snapshotting Kokkos into Trilinos, DynRankView / View interoperability is now available in Trilinos develop branch. Summary:

DynRankView redesigned to be its own View implementation
Added features:
* Copy, Copy-assign View to DynRankView
* deep_copy between Views and DynRankViews
* subview function name now enabled (subdynrankview has same functionality)
* Common free function to request rank of a View and DynRankView:  rank(user_view); 
  This will eventually be replaced by a constexpr rank() in the View, but this will require coordination with packages and apps to avoid breaking their builds.

**Comment by eric-c-cyr** [comment]
Is there a status update on this issue?

**Comment by rppawlo** [comment]
Not finished. All the pieces are in place. We just need to change over to using the refactor branch. Hope to get to this very soon.

**Comment by kyungjoo-kim** [comment]
Since Andy went through a first iteration and convert the code for the refactored version, you would not have much trouble.

**Comment by rppawlo** [comment]
Finally back on this. @kyungjoo-kim and I have been going back and forth on this. Hope to have it finished up this week.

**Comment by rppawlo** [comment]
Currently blocked by required changes in sacado. That is being tracked in #685.

Here is @kyungjoo-kim 's assessment:

Okay. I looked at this problem and found an issue in createDynRankView in View Factory.

When we use subview of a view, it returns strided layout and this layout is inherited to view factory. This is wrong. It has to be either left/right layout. Eric P knows this and he will fix the problem and add unit tests for that. Without that FAC machinery, I cannot make the code correct.

**Comment by rppawlo** [comment]
Updating:  Kyungjoo and I  have made some progress on the integration. Hope to finish off this week or next at the latest.

**Comment by rppawlo** [comment]
Completed and pushed. jenkins CI is passing with the changes.
